offload_data/plugin.cpp

```diff
@@ -0,0 +1,184 @@
+		, _stat_processed(0)
+		, _stat_missed(0)
+		, percent{0}
+		, img_idx{0}
+		, enable_offload{ILLIXR::str_to_bool(ILLIXR::getenv_or("ILLIXR_OFFLOAD_ENABLE", "False"))}
+		, is_success{true}
+		/// TODO: Set with #198
+		, obj_dir{ILLIXR::getenv_or("ILLIXR_OFFLOAD_PATH", "metrics/offloaded_data/")}
+		{
+			/// Remove existing file and create folder for offloading
+			/// TODO: Not portable
+			system(("rm -r " + obj_dir).c_str());
+			system(("mkdir -p " + obj_dir).c_str());
+		}
+
+		virtual skip_option _p_should_skip() override {
+			auto in = _offload_data_reader->get_latest_ro();
+			if (!in || in->seq == _seq_expect-1) {
+				// No new data, sleep
+				std::this_thread::sleep_for(std::chrono::milliseconds{1});
+				return skip_option::skip_and_yield;
+			} else {
+				if (in->seq != _seq_expect) {
+					_stat_missed = in->seq - _seq_expect;
+				} else {
+					_stat_missed = 0;
+				}
+				_stat_processed++;
+				_seq_expect = in->seq+1;
+				return skip_option::run;
+			}
+		}
+
+		void _p_one_iteration() override {
+			std::cout << "Image index: " << img_idx++ << std::endl;
+			_offload_data_container.push_back(_offload_data_reader->get_latest_ro());
+		}
+
+		virtual ~offload_data() override {
+			// Write offloaded data from memory to disk
+			if (enable_offload)
+				writeDataToDisk(_offload_data_container);
+		}
+
+private:
+	const std::shared_ptr<switchboard> sb;
+	std::unique_ptr<reader_latest<texture_pose>> _offload_data_reader;
+	long long _seq_expect, _stat_processed, _stat_missed;
+	std::vector<int> _time_seq;
+	std::vector<const texture_pose*> _offload_data_container;
+
+	int percent;
+	int img_idx;
+	bool enable_offload;
+	bool is_success;
+	std::string obj_dir;
+
+	void writeMetadata(std::vector<int> _time_seq)
+	{
+		double mean  = std::accumulate(_time_seq.begin(), _time_seq.end(), 0.0) / _time_seq.size();
+		double accum = 0.0;
+		std::for_each(std::begin(_time_seq), std::end(_time_seq), [&](const double d){
+			accum += (d - mean) * (d - mean);
+		});
+		double stdev = sqrt(accum/(_time_seq.size() - 1));
+
+		std::vector<int>::iterator max = std::max_element(_time_seq.begin(), _time_seq.end());
+		std::vector<int>::iterator min = std::min_element(_time_seq.begin(), _time_seq.end());
+
+		std::ofstream meta_file (obj_dir + "metadata.out");
+		if (meta_file.is_open())
+		{
+			meta_file << "mean: " << mean << std::endl;
+			meta_file << "max: " << *max << std::endl;
+			meta_file << "min: " << *min << std::endl;
+			meta_file << "stdev: " << stdev << std::endl;
+			meta_file << "total number: " << _time_seq.size() << std::endl;
+
+			meta_file << "raw time: " << std::endl;
+			for (int& it : _time_seq)
+				meta_file << it << " ";
+			meta_file << std::endl << std::endl << std::endl;
+
+			meta_file << "ordered time: " << std::endl;
+			std::sort(_time_seq.begin(), _time_seq.end(), [](int x, int y) {return x > y;});
+			for (int& it : _time_seq)
+				meta_file << it << " ";
+		}
+		meta_file.close();
+	}
```

Comment:
`meta_file.close()` may still fail if something goes wrong. We may print a warning here in that scenario, if desired.